sourcegraph/sourcegraph
# Add "request access" experimental feature

Closes https://github.com/sourcegraph/sourcegraph/issues/47445.

[Demo video](https://www.loom.com/share/8c9dd4d916f8421b96bb9be35dad0960)

## Screenshots

| Description | Screenshot |
| --: | :-- |
| Sign in page when builtin auth provider allowSignup=false | <img width="1724" alt="image" src="https://user-images.githubusercontent.com/6717049/219633483-0bd31e33-2afc-47e7-9727-940b946f52db.png">  |
| Request access form page | <img width="1724" alt="image" src="https://user-images.githubusercontent.com/6717049/219635313-8a086977-d250-4e81-9e1c-6d0db43e20de.png"> |
| Request access error submission page: user with email already exists | <img width="1724" alt="image" src="https://user-images.githubusercontent.com/6717049/219635840-ec636ecf-8b16-463a-b78b-0d9a511e8ed7.png"> |
| Request access error submission page: access request with email already exists | <img width="1724" alt="image" src="https://user-images.githubusercontent.com/6717049/219638764-55eb77d6-6b09-44bb-8af2-9b54acfd0270.png"> |
| Request access submission success page | <img width="1724" alt="image" src="https://user-images.githubusercontent.com/6717049/219636963-62c2b44d-2c37-469a-b8eb-3b1afc5205e0.png"> |
| Site admin global navbar notification | <img width="1728" alt="image" src="https://user-images.githubusercontent.com/6717049/220896058-3805ecbf-ff79-4e6e-8ad2-b4a8b3ae0cbd.png"> |
| Site admin / Access Requests page | <img width="1728" alt="image" src="https://user-images.githubusercontent.com/6717049/220896187-bec260e1-4b92-44a7-96c3-b29ad507a434.png"> |
| Site admin / Access Requests / Approve | <img width="1728" alt="image" src="https://user-images.githubusercontent.com/6717049/220896585-961dfba0-15b7-4e9e-9124-6d313d59c1c8.png"> |
| Site admin / Access Requests / Approve with error |  <img width="1728" alt="image" src="https://user-images.githubusercontent.com/6717049/220896826-a2263dd7-bee3-4ae9-ab10-759ca7de8d40.png"> |

## Test plan
- `sg start`
- Open http://localhost:3080/site-admin/configuration and disable builtin auth provider allowSignup
- Open http://localhost:3080/sign-in in incognito
- Click the "Request access" link
- Fill the form and submit a request
- Go to http://localhost:3080 (as site admin)
- Check there is an "Access requests #" button in the global navbar
- Click to that button
- Approve/Reject access request
